base_local_planner/include/base_local_planner/speed_limiters/speed_limiter.h

```diff
@@ -0,0 +1,99 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of 6 River Systems. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Daniel Grieneisen
+ *********************************************************************/
+
+#ifndef SPEED_LIMITER_H_
+#define SPEED_LIMITER_H_
+
+#include <costmap_2d/costmap_2d_ros.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <tf/tf.h>
+
+namespace base_local_planner {
+
+/**
+ * This class limits the velocity of the robot
+ */
+class SpeedLimiter {
+public:
+  /**
+   * Constructor
+   */
+  SpeedLimiter(costmap_2d::Costmap2DROS* costmap) : costmap_(costmap) {};
+
+  /**
+   * Destructor
+   */
+  virtual ~SpeedLimiter() {}
+
+  virtual void initialize(std::string name) = 0;
+
+  /**
+   * Calculate limits
+   * @return true if preparations were successful
+   */
+  virtual bool calculateLimits(double& max_allowed_linear_vel, double& max_allowed_angular_vel) = 0;
+
+  
+  void setPlan(const std::vector<geometry_msgs::PoseStamped>& plan) {
+    plan_ = plan;
+  };
+
+  void setMaxLimits(double linear, double angular) {
+    max_linear_velocity_ = linear;
+    max_angular_velocity_ = angular;
+  };
+
+protected:
+  std::shared_ptr<tf::Stamped<tf::Pose>> getCurrentPose() {
+    auto robot_pose = std::make_shared<tf::Stamped<tf::Pose>>();
+    if (!costmap_->getRobotPose(*robot_pose)) {
+      ROS_WARN("Could not get robot pose to calculate speed limits");
+      return nullptr;
+    }  
+    return robot_pose;  
+  }
+
+  costmap_2d::Costmap2DROS* costmap_;
```

Comment:
this should be a smart pointer

Reply:
It's passed down through a few layers of classes as a raw pointer.  I don't necessarily want to change the interface at the top level where it is passed in (as this is a library that we have forked).  I'm not sure if I should just cast the raw pointer into a smart pointer, or only use a raw pointer.